examples/chat-in-the-browser/index.js

```diff
@@ -0,0 +1,97 @@
+import Boostrap from 'libp2p-bootstrap'
+import pipe from 'it-pipe'
+
+document.addEventListener('DOMContentLoaded', async () => {
+  // Create our libp2p node
+  const libp2p = await Libp2p.create({
+    modules: {
+      transport: [Websockets, WebRTCStar],
+      connEncryption: [Secio],
+      streamMuxer: [Mplex],
+      peerDiscovery: [Boostrap]
+    },
+    config: {
+      peerDiscovery: {
+        bootstrap: {
+          enabled: true,
+          list: [
+            '/dns4/ams-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLer265NRgSp2LA3dPaeykiS1J6DifTC88f5uVQKNAd',
+            '/dns4/lon-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLMeWqB7YGVLJN3pNLQpmmEk35v6wYtsMGLzSr5QBU3',
+            '/dns4/sfo-3.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLPppuBtQSGwKDZT2M73ULpjvfd3aZ6ha4oFGL1KrGM',
+            '/dns4/sgp-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLSafTMBsPKadTEgaXctDQVcqN88CNLHXMkTNwMKPnu',
+            '/dns4/nyc-1.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLueR4xBeUbY9WZ9xGUUxunbKWcrNFTDAadQJmocnWm',
+            '/dns4/nyc-2.bootstrap.libp2p.io/tcp/443/wss/p2p/QmSoLV4Bbm51jM9C4gDYZQ9Cy3U6aXMJDAbzgu2fzaDs64'
+          ]
+        }
+      }
+    }
+  })
+
+  // UI elements
+  const status = document.getElementById('status')
+  const output = document.getElementById('output')
+  const txtSend = document.getElementById('txt_send')
+
+  output.textContent = ''
+
+  function log (txt) {
+    console.info(txt)
+    output.textContent += `${txt.trim()}\n`
+  }
+
+  // Add the signaling server address, along with our PeerId to our multiaddrs list
+  // libp2p will automatically attempt to dial to the signaling server so that it can
+  // receive inbound connections from other peers
+  const webrtcAddr = '/ip4/0.0.0.0/tcp/9090/wss/p2p-webrtc-star'
+  libp2p.peerInfo.multiaddrs.add(webrtcAddr)
+
+  // Listen for new peers
+  libp2p.on('peer:discovery', peerInfo => {
+    log(`Found peer ${peerInfo.id.toB58String()}`)
+  })
+
+  // Listen for new connections to peers
+  let dialedStream = null
+  libp2p.on('peer:connect', peerInfo => {
+    log(`Connected to ${peerInfo.id.toB58String()}`)
+    libp2p.dialProtocol(peerInfo, ['/chat']).then(({ stream }) => {
+      log('dialed a stream', stream)
+      dialedStream = stream
+    })
+  })
+
+  // Listen for peers disconnecting
+  libp2p.on('peer:disconnect', peerInfo => {
+    log(`Disconnected from ${peerInfo.id.toB58String()}`)
+  })
+
+  await libp2p.start()
+  status.innerText = 'libp2p started!'
+  log(`libp2p id is ${libp2p.peerInfo.id.toB58String()}`)
+
+  let handledStream = null
+  await libp2p.handle(['/chat'], ({ connection, stream }) => {
+    log(`handle chat from ${connection.remotePeer.toB58String()}`)
+    handledStream = stream
+    pipe(handledStream, async function (source) {
+      for await (const msg of source) {
+        log(`Received message: ${msg}`)
+      }
+    })
+  })
+
+  function send () {
+    const value = txtSend.value
+    pipe([value], dialedStream)
```

Comment:
The problem you're having is here, `pipe([value], dialedStream)`. What this does is write `value` and then it closes the write side of this stream, so you **cant** send anymore data. Since the write side of the stream is closing, your `for await` log loop on the receiver side will close the read side of the connection, but not the write. So you have half open streams right now.

You can do 1 of two things.
1. Keep the stream open and pass chats back and forth over it using something like `it-pushable`, or an async generator that yields your messages, so the stream doesn't close unless you explicitly close it or it errors.
2. Just create a new `/chat` stream and send the data when it needs to be sent.

Option 2 is likely going to be safest. There is some minor cost overhead creating the streams, but the connection is still open so it's pretty low. You can also 'ack' messages by sending back an empty value, `pipe([], dialedStream)`, which will fully close the stream.

I'd recommend updating `send` to take a peer of who you want to send to. 

```js
async function send (peerInfo, message) {
  try {
    const { stream } = await libp2p.dialProtocol(peerInfo, ['/chat'])
    await pipe([message], stream, consume) // consume from `streaming-iterables` will wait for the empty 'ack'
  } catch (err) {
    console.log('Could send the message', err)
  }
}
```

Reply:
Yes that confirms my suspicion. I was thinking about a way to accomplish (1) using an async generator, but I wasn't sure how to do that. I'll try with the `it-pushable`.

As for your suggestion (2), I'm not sure what you mean. What API call were you thinking of to create that new stream?

(Thanks for responding so fast by the way!)

Reply:
The `send` function above will do this. When you want to send a message to a certain peer just pass their PeerInfo and the message to the send function. If there is already a connection `dialProtocol` will just create a new stream on the existing connection. It will send the data and then wait for the ack.

Your incoming message handler could then look something like:
```js
await libp2p.handle(['/chat'], async ({ connection, stream }) => {
  log(`handle chat from ${connection.remotePeer.toB58String()}`)
  // Read in all messages directly from the streams `source` async iterable
  for await (const msg of stream.source) {
    log(`Received message: ${msg}`)
  }
  // Write the ack (an empty reply) message to close the stream. You can call sink to do this directly
  await stream.sink([])
  log('Ack sent, stream is closed')
})
```

Note, you can use `pipe` instead of calling the stream methods, but this might be a bit clearer for a read and respond approach.

Reply:
Thanks for clarifying @jacobheun! I made a few more commits. One where the example runs with `it-pushable`, which works. Then I moved on to your option 2 as mentioned above.

> Option 2 is likely going to be safest. There is some minor cost overhead creating the streams, but the connection is still open so it's pretty low. You can also 'ack' messages by sending back an empty value, `pipe([], dialedStream)`, which will fully close the stream.

I updated the example to use the `send` method you suggested, and it works as well. This raises two questions for me:

* You mentioned the `pipe([], dialedStream)` call to close the stream. I assume this was meant to be on the "handle" side of things, so `pipe([], handledStream)`, right? (This is how I now put it in the example.)
* Can you clarify what you meant by saying that this option is "safest"? Safe in what way? What kind of problems have you seen that are solved by this?

Based on your answers I will add some more comments to the example. Thanks again!!

Reply:
> You mentioned the pipe([], dialedStream) call to close the stream. I assume this was meant to be on the "handle" side of things, so pipe([], handledStream), right?

Yes, that was a mistake on my part

> Can you clarify what you meant by saying that this option is "safest"? Safe in what way?

There are 2 things you'll want to make sure to account for.
1. `it-pushable` has no flow control, which means you need to deal with write overflows backoff on the writer side. The nice thing about async iterables is allowing the stream to read in the messages rather than pushing them.
2. Keeping a long lived stream means you need to take better care about closing it when you no longer need it. The nice thing about option 2 is that the stream only lives as long as the message send and ack, so there's less risk in accidentally leaking streams by not closing them.

Reply:
That’s super helpful, thanks Jacob!